File: src/menu.ts
/**
 * 导航菜单的纯逻辑。
 *
 * 菜单是**配置**而不是内容（理由在 `docs/configuration.md`），但它渲染在每一页的头部，
 * 所以一个错地址就是全站死链——而死链只能等「体检」才被发现，那已经是生成之后。
 * 这里两个函数都是为了把问题提前到「编辑菜单的那一刻」。
 *
 * 抽成纯函数是因为判断规则（末尾斜杠、锚点、站外链接）全是边界情况，
 * 留在组件里只能靠肉眼。
 */
import type { LinkTarget, SectionTarget } from './crossref'

/** 菜单项里这里只关心地址。 */
interface UrlOnly {
  readonly url: string
}

/**
 * 比较地址时先规整：**末尾斜杠不算差异**。
 *
 * 产物里 `/about` 与 `/about/` 打开的是同一个 `index.html`（`links.rs` 的死链体检
 * 也是这么宽容的），把它们当两个地址会报出一堆假问题。锚点一并去掉：
 * `/about/#history` 指的还是 `/about/`。
 */
function normalize(url: string): string {
  const withoutHash = url.split('#')[0].trim()
  return withoutHash.length > 1 ? withoutHash.replace(/\/+$/, '') : withoutHash
}

/** 站外地址（它的正确性不在这个站里，管不了也不该管）。 */
function isExternal(url: string): boolean {
  return /^(https?:)?\/\//.test(url) || url.startsWith('mailto:')
}

/**
 * 还没出现在菜单里的栏目。
 *
 * 新建一个栏目之后最容易忘的就是「去导航里加一行」——文档里专门写了这一步，
 * 说明它确实要靠人记。这里把「还缺哪些」算出来，界面可以一键补齐。
 *
 * 没有索引页的栏目不算：那个地址本身是 404（`menuTargets` 同样滤掉它们）。
 */
export function missingSections(
  menu: readonly UrlOnly[],
  sections: readonly SectionTarget[],
): SectionTarget[] {
  const taken = new Set(menu.map((item) => normalize(item.url)))
  return sections.filter(
    (section) => section.index_source && !taken.has(normalize(section.url)),
  )
}

/**
 * 菜单里指向站内、但站内找不到的地址。
 *
 * 只报**站内**的：站外链接与纯锚点的正确性不在这个站里。空地址也不报——那是还没填完的
 * 行，保存时 `MenuItem::validate` 自会拦下，在这里再喊一遍只是噪音。
 *
 * **首页也要真有 `index.md` 才算存在**：没有它，站点根就是 404，报出来是对的。
 * core 侧（`menu.rs`）拿产物地址做同一件事，两边不该有一边格外宽容。
 *
 * 返回的是**去重后**的地址列表：同一个错地址报两遍不会让人更快改对。
 */
export function unknownMenuUrls(
  menu: readonly UrlOnly[],
  targets: readonly LinkTarget[],
): string[] {
  const known = new Set(targets.map((item) => normalize(item.url)))
  const bad = new Set<string>()
  for (const item of menu) {
    const url = item.url.trim()
    if (!url || isExternal(url) || url.startsWith('#')) continue
    if (!known.has(normalize(url))) bad.add(item.url)
  }
  return [...bad]
}

File: crates/staticsmith-core/src/menu.rs
//! 导航菜单的校验。
//!
//! 与前端 `src/menu.ts` 是一对**孪生实现**：同一套规则要在两个时机生效——
//! 编辑菜单时立刻提示（前端，不走 IPC 才谈得上「立刻」），生成时再报一次
//! （这里，CLI 与 MCP 也就都有了）。两边各有测试钉同样的例子，改一边就得改另一边。
//!
//! 为什么值得做两遍：菜单渲染在**每一页**的头部，一个错地址就是全站死链，
//! 而死链原先只能等 [`crate::links`] 的体检发现——那已经是生成之后的事。

use std::collections::BTreeSet;

use crate::config::MenuItem;

/// 比较地址时先规整：**末尾斜杠与锚点都不算差异**。
///
/// 产物里 `/about` 与 `/about/` 打开的是同一个 `index.html`（[`crate::links`] 的死链
/// 体检也是这么宽容的），把它们当两个地址会报出一堆假问题。`/about/#history`
/// 指的还是 `/about/`。
pub fn normalize_url(url: &str) -> &str {
    let without_hash = url.split('#').next().unwrap_or("").trim();
    if without_hash.len() > 1 {
        without_hash.trim_end_matches('/')
    } else {
        without_hash
    }
}

/// 菜单里指向站内、但站内没有这个地址的那些项。
///
/// 只查**站内**的：站外链接与纯锚点的正确性不在这个站里。空地址也不查——
/// 那是 [`MenuItem::validate`] 的活儿，在这里再报一遍只是噪音。
///
/// `known` 收的是**产物地址**（每个已发布页面的 `url`，包括栏目索引页）。
/// 没有索引页的栏目因此自然不在其中——它的地址本来就是 404。
pub fn unknown_urls<'a>(menu: &'a [MenuItem], known: &BTreeSet<String>) -> Vec<&'a str> {
    let mut seen = BTreeSet::new();
    let mut bad = Vec::new();
    for item in menu {
        let url = item.url.trim();
        if url.is_empty() || item.is_external() || url.starts_with('#') {
            continue;
        }
        if !known.contains(normalize_url(url)) && seen.insert(normalize_url(url)) {
            bad.push(item.url.as_str());
        }
    }
    bad
}

/// 这条警告的措辞只有这一份（界面、CLI、MCP 共用同一句）。
///
/// 说清「在哪儿」和「会怎样」：用户看到的是构建报告，而菜单在每一页上，
/// 不点明「每一页」他会以为只是某一处小毛病。
pub fn unknown_urls_warning(bad: &[&str]) -> String {
    format!(
        "导航菜单里有 {} 个地址在站内找不到（{}）——它们出现在每一页的头部，都是死链",
        bad.len(),
        bad.join("、")
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    fn item(url: &str) -> MenuItem {
        MenuItem {
            name: "项".into(),
            url: url.into(),
            weight: 0,
            blank: false,
        }
    }

    fn known() -> BTreeSet<String> {
        ["/", "/posts", "/about"]
            .into_iter()
            .map(String::from)
            .collect()
    }

    #[test]
    fn reports_only_urls_the_site_does_not_have() {
        let menu = vec![item("/posts/"), item("/abuot/")];
        assert_eq!(unknown_urls(&menu, &known()), vec!["/abuot/"]);
    }

    #[test]
    fn external_links_and_anchors_are_not_our_business() {
        let menu = vec![
            item("https://example.com/"),
            item("//cdn.example.com/x"),
            item("mailto:a@b.c"),
            item("#top"),
            item("/about/#history"),
        ];
        assert!(unknown_urls(&menu, &known()).is_empty());
    }

    #[test]
    fn trailing_slash_is_not_a_difference() {
        let menu = vec![item("/about"), item("/about/")];
        assert!(unknown_urls(&menu, &known()).is_empty());
    }

    #[test]
    fn empty_url_is_left_to_validate() {
        let menu = vec![item(""), item("   ")];
        assert!(unknown_urls(&menu, &known()).is_empty());
    }

    #[test]
    fn the_same_bad_url_is_reported_once() {
        let menu = vec![item("/nope/"), item("/nope")];
        assert_eq!(unknown_urls(&menu, &known()), vec!["/nope/"]);
    }

    /// 措辞里必须点明「每一页」，否则会被当成某一处的小毛病。
    #[test]
    fn warning_says_where_it_hurts() {
        let text = unknown_urls_warning(&["/a/", "/b/"]);
        assert!(text.contains("2 个"));
        assert!(text.contains("/a/、/b/"));
        assert!(text.contains("每一页"));
    }
}
